clients/pkg/promtail/scrapeconfig/scrapeconfig.go

```diff
@@ -33,18 +33,21 @@ import (
 
 // Config describes a job to scrape.
 type Config struct {
-	JobName                string                     `yaml:"job_name,omitempty"`
-	PipelineStages         stages.PipelineStages      `yaml:"pipeline_stages,omitempty"`
-	JournalConfig          *JournalTargetConfig       `yaml:"journal,omitempty"`
-	SyslogConfig           *SyslogTargetConfig        `yaml:"syslog,omitempty"`
-	GcplogConfig           *GcplogTargetConfig        `yaml:"gcplog,omitempty"`
-	PushConfig             *PushTargetConfig          `yaml:"loki_push_api,omitempty"`
-	WindowsConfig          *WindowsEventsTargetConfig `yaml:"windows_events,omitempty"`
-	KafkaConfig            *KafkaTargetConfig         `yaml:"kafka,omitempty"`
-	GelfConfig             *GelfTargetConfig          `yaml:"gelf,omitempty"`
-	CloudflareConfig       *CloudflareConfig          `yaml:"cloudflare,omitempty"`
-	RelabelConfigs         []*relabel.Config          `yaml:"relabel_configs,omitempty"`
-	ServiceDiscoveryConfig ServiceDiscoveryConfig     `yaml:",inline"`
+	JobName          string                     `yaml:"job_name,omitempty"`
+	PipelineStages   stages.PipelineStages      `yaml:"pipeline_stages,omitempty"`
+	JournalConfig    *JournalTargetConfig       `yaml:"journal,omitempty"`
+	SyslogConfig     *SyslogTargetConfig        `yaml:"syslog,omitempty"`
+	GcplogConfig     *GcplogTargetConfig        `yaml:"gcplog,omitempty"`
+	PushConfig       *PushTargetConfig          `yaml:"loki_push_api,omitempty"`
+	WindowsConfig    *WindowsEventsTargetConfig `yaml:"windows_events,omitempty"`
+	KafkaConfig      *KafkaTargetConfig         `yaml:"kafka,omitempty"`
+	GelfConfig       *GelfTargetConfig          `yaml:"gelf,omitempty"`
+	CloudflareConfig *CloudflareConfig          `yaml:"cloudflare,omitempty"`
+	DockerConfig     *DockerConfig              `yaml:"docker,omitempty"`
+	RelabelConfigs   []*relabel.Config          `yaml:"relabel_configs,omitempty"`
+	// List of Docker service discovery configurations.
+	DockerSDConfigs        []*moby.DockerSDConfig `yaml:"docker_sd_configs,omitempty"`
```

Comment:
Why not add this into the `ServiceDiscoveryConfig` struct ? I guess that doesn't make much difference but at least it's aligned with the rest.

Reply:
It cannot be part of the `ServiceDiscoveryConfig` because [HasServiceDiscoveryConfig](https://github.com/grafana/loki/blob/main/clients/pkg/promtail/scrapeconfig/scrapeconfig.go#L370) would return true. This would trigger a `NewFileTargetManager` which we are not using for Docker. I also remember that the unused file target manager would have a nil pointer panic down the line somewhere.